desc/compute/_neoclassical.py

```diff
@@ -0,0 +1,626 @@
+from ..utils import cross, dot, safediv
+from .data_index import register_compute_fun
+
+_bounce_doc = {
+    "theta": """jnp.ndarray :
+        Shape (num rho, X, Y).
+        DESC coordinates θ sourced from the Clebsch coordinates
+        ``FourierChebyshevSeries.nodes(X,Y,rho,domain=(0,2*jnp.pi))``.
+        Use the ``Bounce2D.compute_theta`` method to obtain this.
+        """,
+    "Y_B": """int :
+        Desired resolution for |B| along field lines to compute bounce points.
+        Default is double ``Y``.
+        """,
+    "num_transit": """int :
+        Number of toroidal transits to follow field line.
+        For axisymmetric devices, one poloidal transit is sufficient. Otherwise,
+        assuming the surface is not near rational, more transits will
+        approximate surface averages better, with diminishing returns.
+        """,
+    "num_well": """int :
+        Maximum number of wells to detect for each pitch and field line.
+        Giving ``None`` will detect all wells but due to current limitations in
+        JAX this will have worse performance.
+        Specifying a number that tightly upper bounds the number of wells will
+        increase performance. In general, an upper bound on the number of wells
+        per toroidal transit is ``Aι+B`` where ``A``,``B`` are the poloidal and
+        toroidal Fourier resolution of |B|, respectively, in straight-field line
+        PEST coordinates, and ι is the rotational transform normalized by 2π.
+        A tighter upper bound than ``num_well=(Aι+B)*num_transit`` is preferable.
+        The ``check_points`` or ``plot`` methods in ``desc.integrals.Bounce2D``
+        are useful to select a reasonable value.
+        """,
+    "num_quad": """int :
+        Resolution for quadrature of bounce integrals.
+        Default is 32. This parameter is ignored if given ``quad``.
+        """,
+    "num_pitch": "int : Resolution for quadrature over velocity coordinate.",
+    "batch_size": """int :
+        Number of pitch values with which to compute simultaneously.
+        If given ``None``, then ``batch_size`` defaults to ``num_pitch``.
+        """,
+    "fieldline_quad": """tuple[jnp.ndarray] :
+        Used to compute the proper length of the field line ∫ dℓ / |B|.
+        Quadrature points xₖ and weights wₖ for the
+        approximate evaluation of the integral ∫₋₁¹ f(x) dx ≈ ∑ₖ wₖ f(xₖ).
+        Default is Gauss-Legendre quadrature at resolution ``Y_B//2``
+        on each toroidal transit.
+        """,
+    "quad": """tuple[jnp.ndarray] :
+        Used to compute bounce integrals.
+        Quadrature points xₖ and weights wₖ for the
+        approximate evaluation of the integral ∫₋₁¹ f(x) dx ≈ ∑ₖ wₖ f(xₖ).
+        """,
+    "spline": "bool : Whether to use cubic splines to compute bounce points.",
+}
+
+
+def _compute(fun, fun_data, data, theta, grid, num_pitch, simp=False):
+    """Compute ``fun`` for each ρ value iteratively to reduce memory usage.
```

Comment:
Probably linear in number of flux surface in compute. For reverse mode you could remove the call to imap to vectorize and it would probably make less difference on memory since the jacobian computation still dominates.

Reply:
i added batching there too.